File: Backend/src/Fibertest30.Api/Services/RtuTreeService.cs
using Grpc.Core;
using Iit.Fibertest.Dto;
using MediatR;

namespace Fibertest30.Api;

public class RtuTreeService : RtuTree.RtuTreeBase
{
    private readonly ISender _mediator;

    public RtuTreeService(ISender mediator)
    {
        _mediator = mediator;
    }

    public override async Task<GetRtuTreeResponse> GetRtuTree(GetRtuTreeRequest request, ServerCallContext context)
    {
        var tree = await _mediator.Send(new GetRtuTreeQuery(), context.CancellationToken);
        return new GetRtuTreeResponse() { Rtus = { tree.Select(r => r.ToProto()) } };
    }

    public override async Task<GetRtuResponse> GetRtu(GetRtuRequest request, ServerCallContext context)
    {
        var rtuDto = await _mediator.Send(new GetRtuQuery(request.RtuId), context.CancellationToken);
        return new GetRtuResponse() { Rtu = rtuDto.ToProto() };
    }

    public override async Task<AttachTraceResponse> AttachTrace(AttachTraceRequest request, ServerCallContext context)
    {
        var dto = new AttachTraceDto { TraceId = Guid.Parse(request.TraceId) };
        var ports = new List<OtauPortDto>(request.PortOfOtau.Select(p => p.FromProto()));
        dto.OtauPortDto = ports[0];
        if (ports.Count > 1)
            dto.MainOtauPortDto = ports[1];

        await _mediator.Send(new AttachTraceCommand(dto), context.CancellationToken);
        return new AttachTraceResponse();
    }

    public override async Task<DetachTraceResponse> DetachTrace(DetachTraceRequest request, ServerCallContext context)
    {
        var guid = Guid.Parse(request.TraceId);
        await _mediator.Send(new DetachTraceCommand(guid), context.CancellationToken);
        return new DetachTraceResponse();
    }

    public override async Task<AttachOtauResponse> AttachOtau(AttachOtauRequest request, ServerCallContext context)
    {
        var dto = request.FromProto();
        await _mediator.Send(new AttachOtauCommand(dto), context.CancellationToken);
        return new AttachOtauResponse();
    }

    public override async Task<DetachOtauResponse> DetachOtau(DetachOtauRequest request, ServerCallContext context)
    {
        var dto = request.FromProto();
        await _mediator.Send(new DetachOtauCommand(dto), context.CancellationToken);
        return new DetachOtauResponse();
    }

    public override async Task<GetTraceBaselineStatResponse> GetTraceBaselineStat(GetTraceBaselineStatRequest request,
        ServerCallContext context)
    {
        var baselines = await _mediator.Send(new GetTraceBaselineStatQuery(Guid.Parse(request.TraceId)),
            context.CancellationToken);
        return new GetTraceBaselineStatResponse() { Baselines = { baselines.Select(b => b.ToProto()) } };
    }

    public override async Task<GetTraceStatisticsResponse> GetTraceStatistics(GetTraceStatisticsRequest request,
        ServerCallContext context)
    {
        var measurements = await _mediator.Send(new GetTraceMeasurementStatQuery(Guid.Parse(request.TraceId)),
            context.CancellationToken);
        return new GetTraceStatisticsResponse() { Measurements = { measurements.Select(m => m.ToProto()) } };
    }

    public override async Task<GetTraceLastMeasurementResponse> GetTraceLastMeasurement(
        GetTraceLastMeasurementRequest request, ServerCallContext context)
    {
        var sorFileId = await _mediator.Send(new GetTraceLastMeasurement(Guid.Parse(request.TraceId)),
            context.CancellationToken);

        return new GetTraceLastMeasurementResponse() { SorFileId = sorFileId };
    }
}

File: WebFt30/src/app/core/grpc/services/rtu-tree.service.ts
import { Injectable, inject } from '@angular/core';
import { GrpcWebFetchTransport } from '@protobuf-ts/grpcweb-transport';
import * as grpc from 'src/grpc-generated';
import { AuthInterceptor } from '../auth.interceptor';
import { GrpcUtils } from '../grpc.utils';
import { Observable } from 'rxjs';
import { AttachTraceDto } from '../../store/models/ft30/attach-trace-dto';
import { FtBaseMapping } from '../../store/mapping/ft-base-mapping';
import { AttachOtauDto } from '../../store/models/ft30/attach-otau-dto';
import { DetachOtauDto } from '../../store/models/ft30/detach-otau-dto';

@Injectable({
  providedIn: 'root'
})
export class RtuTreeService {
  private authInterceptor = inject(AuthInterceptor);
  private client: grpc.RtuTreeClient;

  constructor() {
    const interceptors = [this.authInterceptor.toRpcInterceptor()];
    const options = GrpcUtils.getGrpcOptions(GrpcUtils.getApiUrl(), interceptors);
    const transport = new GrpcWebFetchTransport(options);
    this.client = new grpc.RtuTreeClient(transport);
  }

  refreshRtuTree(): Observable<grpc.GetRtuTreeResponse> {
    const request: grpc.GetRtuTreeRequest = {};
    return GrpcUtils.unaryToObservable(this.client.getRtuTree.bind(this.client), request, {});
  }

  getOneRtu(rtuId: string): Observable<grpc.GetRtuResponse> {
    const request: grpc.GetRtuRequest = { rtuId: rtuId };
    return GrpcUtils.unaryToObservable(this.client.getRtu.bind(this.client), request, {});
  }

  attachTrace(dto: AttachTraceDto): Observable<grpc.AttachTraceResponse> {
    const request: grpc.AttachTraceRequest = {
      traceId: dto.traceId,
      portOfOtau: dto.portOfOtau.map((p) => FtBaseMapping.toGrpcPortOfOtau(p))
    };
    return GrpcUtils.unaryToObservable(this.client.attachTrace.bind(this.client), request, {});
  }

  detachTrace(traceId: string): Observable<grpc.DetachTraceResponse> {
    const request: grpc.DetachTraceRequest = { traceId };
    return GrpcUtils.unaryToObservable(this.client.detachTrace.bind(this.client), request, {});
  }

  attachOtau(dto: AttachOtauDto): Observable<grpc.AttachOtauResponse> {
    const request: grpc.AttachOtauRequest = {
      rtuId: dto.rtuId,
      netAddress: FtBaseMapping.toGrpcNetAddress(dto.netAddress),
      opticalPort: dto.opticalPort
    };
    return GrpcUtils.unaryToObservable(this.client.attachOtau.bind(this.client), request, {});
  }

  detachOtau(dto: DetachOtauDto): Observable<grpc.DetachOtauResponse> {
    const request: grpc.DetachOtauRequest = {
      rtuId: dto.rtuId,
      otauId: dto.otauId,
      netAddress: FtBaseMapping.toGrpcNetAddress(dto.netAddress),
      opticalPort: dto.opticalPort
    };
    return GrpcUtils.unaryToObservable(this.client.detachOtau.bind(this.client), request, {});
  }

  getTraceBaselineState(traceId: string): Observable<grpc.GetTraceBaselineStatResponse> {
    const request: grpc.GetTraceBaselineStatRequest = { traceId };
    return GrpcUtils.unaryToObservable(
      this.client.getTraceBaselineStat.bind(this.client),
      request,
      {}
    );
  }

  getTraceStatistics(traceId: string): Observable<grpc.GetTraceStatisticsResponse> {
    const request: grpc.GetTraceStatisticsRequest = { traceId };
    return GrpcUtils.unaryToObservable(
      this.client.getTraceStatistics.bind(this.client),
      request,
      {}
    );
  }

  getTraceLastMeasurement(traceId: string): Observable<grpc.GetTraceLastMeasurementResponse> {
    const request: grpc.GetTraceLastMeasurementRequest = { traceId };
    return GrpcUtils.unaryToObservable(
      this.client.getTraceLastMeasurement.bind(this.client),
      request,
      {}
    );
  }

  getRtuCurrentStep(rtuId: string): Observable<grpc.GetRtuCurrentStepResponse> {
    const request: grpc.GetRtuCurrentStepRequest = { rtuId };
    return GrpcUtils.unaryToObservable(
      this.client.getRtuCurrentStep.bind(this.client),
      request,
      {}
    );
  }
}
